flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -281,10 +289,41 @@ protected void processInput(DefaultActionContext context) throws Exception {
 		if (status == InputStatus.END_OF_INPUT) {
 			context.allActionsCompleted();
 		}
-		else if (status == InputStatus.NOTHING_AVAILABLE) {
+
+		CompletableFuture<?> jointFuture = getInputOutputJointFuture(status);
+		if (jointFuture != null) {
 			SuspendedMailboxDefaultAction suspendedDefaultAction = context.suspendDefaultAction();
-			inputProcessor.getAvailableFuture().thenRun(suspendedDefaultAction::resume);
+			jointFuture.thenRun(suspendedDefaultAction::resume);
+		}
+	}
+
+	/**
+	 * @return a combination of input and output futures if at-least one future of them is not
+	 * completed, otherwise return null if all input and outputs are available.
+	 */
+	private CompletableFuture<?> getInputOutputJointFuture(InputStatus status) {
+		if (status == InputStatus.MORE_AVAILABLE && isOutputAvailable()) {
+			return null;
+		}
+
+		int length = recordWriters.size();
+		for (int i = 0; i < length; i++) {
+			inputOutputFutures[i] = recordWriters.get(i).getAvailableFuture();
+		}
+		inputOutputFutures[length] = inputProcessor.getAvailableFuture();
+		return CompletableFuture.allOf(inputOutputFutures);
+	}
+
+	/**
+	 * @return true if all the record writers are available.
+	 */
+	private boolean isOutputAvailable() {
+		for (RecordWriter recordWriter : recordWriters) {
+			if (!recordWriter.isAvailable()) {
```

Comment:
I ever considered dropping the condition of `isBlocking` before. But considering that `LocalBufferPool#isAvailable` is actually not used in input side, so I retain this condition to not impact the input side. 

After thinking through it now, it is not very clean because during `LocalBufferPool#recycle` we could not distinguish whether it is for input or output usages. 

We can remove this condition here and also adjust the interaction between `RemoteInputChannel` and `LocalBufferPool` to make use of `isAvailable` way instead of `BufferListener` interface. WDYT?

Concerning of the synchronised `recordWriter.isAvailable()`, it is actually the same case as other previous usages. The unavailable state is always visible for mailbox/task thread because the `LocalBufferPool` might become unavailable only after task thread requesting. For the visibility from unavailable to available, it is caused by other threads, then it can be still visible for task thread via touching volatile variable in `CompleteFuture.complete` as we confirmed before.

Reply:
> considering that LocalBufferPool#isAvailable is actually not used in input side

Do you mean that `isBlocking = false` happens only on the inputs? In that case I would say this is a still a buggy code path, that just happens to be not used... at the moment. I would still prefer to have it working in case of some refactorings in the future.

Is there any downside of doing `availabilityHelper.resetUnavailable();` always - including for non blocking cases? Performance overhead?

> We can remove this condition here and also adjust the interaction between RemoteInputChannel and LocalBufferPool to make use of isAvailable way instead of BufferListener interface. 

I think we could do that to unify/simplify things a bit. However I would try to not do this in this PR as I don't think it has that high priority right now. We could create a ticket and log it for the future for someone to grab it.

Reply:
> Do you mean that isBlocking = false happens only on the inputs?

Exactly, the local pool provides two ways of blocking/non-blocking request atm. For the output it uses the blocking way before, and for the input it uses the non-blocking way.

> Is there any downside of doing availabilityHelper.resetUnavailable(); always - including for non blocking cases? Performance overhead?

I am not quite sure of it, so my previous thought is to avoid impacting the unrelevant path in best-effort way, otherwise we have to verify the performance concern.

> I think we could do that to unify/simplify things a bit.....

Agree, if we can convince the trivial impact, the unification work has very low priority and can be focused in future if have time.

Reply:
> I am not quite sure of it, so my previous thought was to avoid impacting the unrelevant path in best-effort way, otherwise we have to verify the performance concern.

I think we have the benchmarks coverage for this, so I would do it and just check the benchmark results.